File: src/vscode-extensions/microsoft-powerplatformlang-extension/client/src/services/telemetryDimensions.ts
/**
 * Canonical telemetry dimension names used across the VS Code extension.
 * All names use camelCase to align with Application Insights conventions.
 * New dimensions MUST be added here before use in code.
 */

// ─────────────────────────────────────────────────────────────────────────────
// Shared dimensions (added automatically to every event by the logger)
// ─────────────────────────────────────────────────────────────────────────────

export const SharedDimensions = {
  /** Session correlation UUID shared between FE and BE. */
  sessionId: 'sessionId',

  /** VSIX version string. */
  version: 'version',

  /** "true" when running in dev/debug mode. Filter: where customDimensions.isDevMode != "true" */
  isDevMode: 'isDevMode',

  /** Event severity: "info" | "warning" | "error". */
  logLevel: 'logLevel',

  /** Human-readable display message. */
  message: 'message',

  /** Error message text (PII-wrapped for scrubbing). */
  errorMessage: 'errorMessage',

  /** Connected agent unique ID. */
  agentId: 'agentId',

  /** Dataverse environment unique ID. */
  environmentId: 'environmentId',

  /** Operation duration in milliseconds. */
  durationMs: 'durationMs',
} as const;

// ─────────────────────────────────────────────────────────────────────────────
// Event-specific dimensions (passed per call site via data parameter)
// ─────────────────────────────────────────────────────────────────────────────

export const EventDimensions = {
  /** User-facing sync command: Preview | Get | Apply. */
  syncOperation: 'syncOperation',

  /** Environment SKU (e.g., Developer, Trial, Sandbox). */
  sku: 'sku',

  /** Number of environments loaded. */
  environmentCount: 'environmentCount',

  /** Number of agents loaded. */
  agentCount: 'agentCount',

  /** LSP method name (e.g., powerplatformls::getWorkspaceDetails). Uses :: to avoid PII flagging. */
  lspMethod: 'lspMethod',
} as const;

File: src/LanguageServers/PowerPlatformLS/Contracts.Internal/Common/TelemetryDimensions.cs
namespace Microsoft.PowerPlatformLS.Contracts.Internal.Common
{
    /// <summary>
    /// Canonical telemetry dimension names used across the language server.
    /// All names use camelCase to align with Application Insights conventions.
    /// New dimensions MUST be added here before use in code.
    /// </summary>
    public static class TelemetryDimensions
    {
        // ─────────────────────────────────────────────────────────────────────
        // Shared dimensions (added to every event)
        // ─────────────────────────────────────────────────────────────────────

        /// <summary>Session correlation UUID shared between FE and BE.</summary>
        public const string SessionId = "sessionId";

        /// <summary>Language server process ID.</summary>
        public const string Pid = "pid";

        /// <summary>VSIX version string.</summary>
        public const string Version = "version";

        /// <summary>Git commit hash of the build.</summary>
        public const string GitHash = "gitHash";

        /// <summary>
        /// "true" when running in dev/debug mode. Added to every event so queries
        /// can filter out dev noise: <c>where customDimensions.isDevMode != "true"</c>
        /// </summary>
        public const string IsDevMode = "isDevMode";

        /// <summary>Connected agent unique ID.</summary>
        public const string AgentId = "agentId";

        /// <summary>Dataverse environment unique ID.</summary>
        public const string EnvironmentId = "environmentId";

        /// <summary>Operation duration in milliseconds.</summary>
        public const string DurationMs = "durationMs";

        // ─────────────────────────────────────────────────────────────────────
        // LSP request dimensions (per handler invocation)
        // ─────────────────────────────────────────────────────────────────────

        /// <summary>Internal LSP request correlation ID. Useful for filtering a session to diagnose one request.</summary>
        public const string ReqId = "reqId";

        /// <summary>LSP method name (e.g., powerplatformls/cloneAgent, textDocument/completion).</summary>
        public const string LspMethod = "lspMethod";

        /// <summary>Handler/operation result: completed | failed | canceled.</summary>
        public const string Outcome = "outcome";

        // ─────────────────────────────────────────────────────────────────────
        // HTTP request dimensions (per HTTP call in LoggingHttpHandler)
        // ─────────────────────────────────────────────────────────────────────

        /// <summary>HTTP verb: GET | POST | PATCH | DELETE.</summary>
        public const string HttpMethod = "httpMethod";

        /// <summary>Normalized URL path with GUIDs replaced by {id}. No query string.</summary>
        public const string HttpEndpoint = "httpEndpoint";

        /// <summary>HTTP response status code (e.g., 200, 401, 500).</summary>
        public const string HttpStatusCode = "httpStatusCode";

        /// <summary>HTTP reason phrase on non-2xx responses (e.g., "Forbidden", "Not Found").</summary>
        public const string HttpReason = "httpReason";

        // ─────────────────────────────────────────────────────────────────────
        // Sync operation dimensions
        // ─────────────────────────────────────────────────────────────────────

        /// <summary>Internal sync function being executed (e.g., FetchRemoteAgent, WriteWorkspace).</summary>
        public const string SyncFunction = "syncFunction";

        // ─────────────────────────────────────────────────────────────────────
        // Error/diagnostic dimensions
        // ─────────────────────────────────────────────────────────────────────

        /// <summary>Error message from failed handler or exception. Scrubbed by PiiScrubTelemetryInitializer.</summary>
        public const string ErrorMessage = "errorMessage";
    }
}
